### PR TITLE
[MODCONSKC-18] - Implement sharing of authorization policies

### PR DESCRIPTION
**Pull request by azizbekxm**
## Purpose
https://folio-org.atlassian.net/browse/MODCONSKC-18
## Approach

- System user is used to apply sharing policies in all tenants, so to perform sharing action it is enough permission to share policies in central tenant
- Existing policy created in some member tenant outside of Consortium manager can not be shared(the same behavior as for Settings)
- New table sharing_policy be created
- SharingSettingService is renamed to some generic name, and it will be used as base class because the same logic will be applied for sharing policies
- SharingPolicyService overriden method to access data in the correct sharing_policy table
- SharingPolicyService share associated role and related role_capabilities, role_capability_sets etc to all member tenants, separate story to do this is created: 
- SharingPolicyService NOT share associated users to all member tenants
- SharingPolicyService set source = ‘CONSORTIUM’ for newly created shared policy
- SharingPolicyController needs to be created

### CODE REVIEW

File: src/main/java/org/folio/consortia/domain/entity/SharingPolicyEntity.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+@Entity
+@Table(name = "sharing_policy")
+public class SharingPolicyEntity extends AuditableEntity {
+  @Id
+  private UUID id;
+  private UUID policyId;
+  private String tenantId;
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (dmtkachenko) [inline src/main/java/org/folio/consortia/domain/entity/SharingPolicyEntity.java:29]: Lombok annotation @EqualsAndHashCode will generate the same code. Any reason for not using it?

File: src/main/resources/db/changelog/changes/create-sharing-policy-table.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                   https://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
+
+  <changeSet id="MODCONSKC-18@@Implement sharing of authorization policies" author="azizbekxm">
+
+    <createTable tableName="sharing_policy">
+      <column name="id" type="uuid">
+        <constraints primaryKey="true" primaryKeyName="pk_shared_policy_id"/>
+      </column>
+      <column name="policy_id" type="uuid"/>
```
Comment (dmtkachenko) [inline src/main/resources/db/changelog/changes/create-sharing-policy-table.xml:13]: Could the field be null?

File: src/main/resources/db/changelog/changes/create-sharing-policy-table.xml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      <column name="tenant_id" type="text"/>
```
Comment (dmtkachenko) [inline src/main/resources/db/changelog/changes/create-sharing-policy-table.xml:14]: Could the field be null?

File: src/main/java/org/folio/consortia/domain/dto/SourceValues.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+public enum SourceValues {
+  CONSORTIUM_FOLIO_INSTANCE("CONSORTIUM-FOLIO"),
+  CONSORTIUM_MARC_INSTANCE("CONSORTIUM-MARC"),
+  CONSORTIUM("consortium"),
+  USER("user");
+
+  private final String value;
+
+  SourceValues(String value) {
```
Comment (BKadirkhodjaev) [inline src/main/java/org/folio/consortia/domain/dto/SourceValues.java:11]: The constructor and the getter on the Enum can replaced with `@Getter` and `@AllArgsConstructor` annotations

**Review by BKadirkhodjaev — APPROVED** [review]

File: src/main/java/org/folio/consortia/domain/entity/SharingPolicyEntity.java (hunk 1 above)
Comment (azizbekxm) [reply src/main/java/org/folio/consortia/domain/entity/SharingPolicyEntity.java:29]: Not big deal, but there is one small reason for this. @EqualsAndHashCode has two approach @EqualsAndHashCode(callSuper=false), @EqualsAndHashCode(callSuper=true). Most entitties extended from AuditableEntity class, and usually ideas warn @EqualsAndHashCode annotation. There might be case somone wrongly configure @EqualsAndHashCode(callSuper=true) cases which compare with superClass as well that is incorrect. Manual implementation as a standart approach chosen for consortia.

File: src/main/resources/db/changelog/changes/create-sharing-policy-table.xml (hunk 3 above)
Comment (azizbekxm) [reply src/main/resources/db/changelog/changes/create-sharing-policy-table.xml]: can't be, and added nullable validation

File: src/main/resources/db/changelog/changes/create-sharing-policy-table.xml (hunk 2 above)
Comment (azizbekxm) [reply src/main/resources/db/changelog/changes/create-sharing-policy-table.xml]: Thanks to point out, can't be, and added nullable validation

File: src/main/java/org/folio/consortia/domain/dto/SourceValues.java (hunk 4 above)
Comment (azizbekxm) [reply src/main/java/org/folio/consortia/domain/dto/SourceValues.java]: Thanks to suggestion, improved it

File: src/main/java/org/folio/consortia/domain/entity/SharingPolicyEntity.java (hunk 1 above)
Comment (azizbekxm) [reply src/main/java/org/folio/consortia/domain/entity/SharingPolicyEntity.java:29]: Overall, using annotation also not bad approach, but need to implement for other entities as well.

**Review by Saba-Zedginidze-EPAM — APPROVED** [review]

File: src/main/java/org/folio/consortia/service/BaseSharingService.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+  private void updateConfigsForFailedTenants(UUID consortiumId, UUID publicationId,
+                                             TRequest sharingConfigRequest) throws InterruptedException {
+    log.debug("updateConfigsForFailedTenants:: Trying to update {}s for failed tenants for consortiumId={}" +
+        " publicationId={} and sharingConfigRequestId={}", getClassName(sharingConfigRequest), consortiumId,
+      publicationId, getConfigId(sharingConfigRequest));
+    boolean isPublicationStatusReady = false;
+    int i = 0;
+    while (Boolean.FALSE.equals(isPublicationStatusReady) && i++ < maxTries) {
+      Thread.sleep(interval);
```
[the remark this answers is not in the document]
Comment (azizbekxm) [reply src/main/java/org/folio/consortia/service/BaseSharingService.java:236]: Good idea, retry is build-in fuction is better to use over manually specified function. changed implementation to use RetryMethod

**Review by OleksiiKuzminov — APPROVED** [review]

**Pull request merged**